pallets/utxo/src/staking.rs

```diff
@@ -0,0 +1,151 @@
+use sp_core::H256;
+use frame_support::{dispatch::{DispatchResultWithPostInfo, Vec}, ensure};
+
+/// A helper trait to handle staking NOT found in pallet-utxo.
+pub trait StakingHelper<AccountId>{
+
+    fn get_account_id(pub_key: &H256) -> AccountId;
+
+    /// start the staking.
+    /// # Arguments
+    /// * `stash_account` - A placeholder of the "supposed" validator. This is only to "satisfy"
+    /// the `pallet-staking`'s needs to be able to stake.
+    /// * `controller_account` - The ACTUAL validator. But this is NOT SO, in the `pallet-staking`.
+    /// In `pallet-staking`, its job is like an "accountant" to the stash account.
+    /// * `rotate_keys` - or also called the session keys, to get up-to-date
+    /// with validators, eras, sessions. see `pallet-session`.
+    fn stake(stash_account:&AccountId, controller_account:&AccountId, rotate_keys:&mut Vec<u8>) -> DispatchResultWithPostInfo;
+
+    /// quitting the role of a validator.
+    fn pause(controller_account:&AccountId) -> DispatchResultWithPostInfo;
+
+    /// transfer balance from the locked state to the actual free balance.
+    fn withdraw(controller_account: &AccountId) -> DispatchResultWithPostInfo;
+
+}
+
+/// performs the staking outside of the `pallet-utxo`.
+pub(crate) fn stake<T: Config>(
+    stash_account: &T::AccountId,
+    controller_account: &T::AccountId,
+    rotate_keys:&Vec<u8>)
+    -> DispatchResultWithPostInfo {
+    if <StakingCount<T>>::contains_key(controller_account) {
+        Err(Error::<T>::StakingAlreadyExists)?
+    }
+
+    let mut rotate_keys = rotate_keys.to_vec();
+    T::StakingHelper::stake(stash_account, controller_account, &mut rotate_keys)?;
+
+    Ok(().into())
+}
+
+/// unlocking the staked funds outside of the `pallet-utxo`.
+/// also means quitting/pausing from being a validator.
+pub(crate) fn unlock<T: Config>(controller_pubkey: &H256) -> DispatchResultWithPostInfo {
+    let controller_account= T::StakingHelper::get_account_id(controller_pubkey);
+    T::StakingHelper::pause(&controller_account)?;
+
+    Ok(().into())
+}
+
+
+/// Consolidates all unlocked utxos  into one, and moves it to `UtxoStore`.
+/// Make SURE that `fn unlock(...)` has been called and the era for withdrawal has passed, before
+/// performing a withdrawal.
+/// # Arguments
+/// * `output` - the output containing the `Destination::WithdrawStake`.
+/// * `hash` - is the outpoint/key to save the new utxo
+pub(crate) fn withdraw<T: Config>(output:&TransactionOutput<T::AccountId>,hash:H256) -> DispatchResultWithPostInfo {
+    if let Destination::WithdrawStake { outpoints, pub_key } =  &output.destination {
+        let controller_account = T::StakingHelper::get_account_id(pub_key);
+        T::StakingHelper::withdraw(&controller_account)?;
+
+        let mut sum:Value = 0;
+        outpoints.iter().for_each(|hash| sum += <LockedUtxos<T>>::take(hash).unwrap().value);
```

Comment:
Is `hash` always going to be in `LockedUtxos` or could the code possibly panic here?